lib/stencil-push.js

```diff
@@ -0,0 +1,242 @@
+const internals = {};
+const Inquirer = require('inquirer');
+
+module.exports = stencilPush;
+module.exports.internals = internals; //TODO: only for testing env 
+
+// ====================================== public =======================================
+function stencilPush(dotStencilFilePath, options, callback) {
+    options = options || {};
+    async.waterfall([
+        async.constant(dotStencilFilePath),
+        internals.readStencilConfigFile,
+        internals.getStoreHash,
+        function(config, cb) {
+            async.parallel({
+                bundlePath: internals.generateBundle.bind(this, config),
+                signature: internals.generateUploadSignature.bind(this, config),
+            }, function(err, results) {
+                console.log('results: ', results);
+                cb(null, results);
+            })
+        },
+        internals.uploadBundle,
+        internals.requestToProcessThemeUpload,
+        async.retryable({
+            errorFilter: function(err) {
+                console.log('=== err: ', err);
+                return true;
+            },
+            times: options.retryTimes === 0 ? 0 : options.retryTimes || 10,
+            interval: function(retryCount) {
+                console.log('=== retryCount: ', retryCount);
+                return 50 * Math.pow(2, retryCount);
+            }
+        }, internals.pollForJobCompletion),
+        internals.promptUserWhetherToApplyTheme,
+        internals.requestVariations,
+        internals.promptUserForVariation,
+        internals.requestToApplyVariation,
+        internals.notifyUserOfCompletion,
+    ], callback)
+}
+
+// ====================================== private ======================================
+
+internals.readStencilConfigFile = function(dotStencilFilePath, callback) {
+    fs.readFile(dotStencilFilePath, { encoding: 'utf8' }, (err, data) => {
+        if (err) {
+            return callback(err);
+        }
+        callback(null, JSON.parse(data));
+    });
+}
+
+internals.getStoreHash = function(config, callback) {
+    Wreck.request('GET', `${config.normalStoreUrl}/admin/oauth/info`, null, (error, response) => {
+        if (error) {
+            return callback(error);
+        }
+
+        if (response.statusCode !== 200) {
+            return callback(new Error('Failed to retrieve store hash'));
+        }
+
+        Wreck.read(response, { json: true }, (err, body) => {
+            if (err) {
+                return callback(err);
+            }
+
+            if (!body.store_hash) {
+                return callback(new Error('Store hash is empty'));
+            }
+
+            callback(null, Object.assign({}, config, { storeHash: body.store_hash }));
+        });
+    });
+}
+
+internals.generateBundle = function(config, callback) {
+    const spawn = require('child_process').spawn;
+    const bundle = spawn(path.resolve(__dirname + '/../bin/stencil'), ['bundle']);
```

Comment:
hmmm..  i didn't see a way to do that without duplicating the code from bin/stencil-bundle so i figured i might as well invoke the script.  but that makes me think, shouldn't we permit the options that are available to stencil bundle?  dest, name, and marketplace?